### ISSUE TITLE
`App` module added by `Content Managers` role doesn't allow `Choose App`

### ISSUE THREAD

**Issue by david-poindexter** [2sic/2sxc#2655]
**I'm submitting a ...** 
[x] bug report

**...about** 
[x] edit experience / UI
[x] admin experience UI
[x] app permissions
[x] other / unknown

**Current behavior** 
We are experiencing this issue on a net new site.

Here is our scenario:
* DNN 9.10.2
* 2sxc 12.8.1
* Have a role called "Content Managers" and they have "Edit" rights on pages.
* Have applied "Can Deploy" to "Content Managers" for "App" extension.

A user belonging to `Content Managers` is able to add an "App" module to a page. However, they do not have the ability to `Choose App` in 2sxc.

In 2sxc, I have tried enabling the Feature for `Permission by Group / Role`. Then for the relevant app, I updated `App Permissions` to add a `Security Rule` for the appropriate RoleId (the one that matches the `Content Managers` role) and gave it `Edit (Create, Read, Update, Delete)` rights. It still didn't work, so I even tried giving it `Full Control (usually host only)`. That didn't work either.

Any other ideas? What the heck am I missing?

I found this SO post helpful, but unfortunately, it just doesn't work for me. ☹️ 
https://stackoverflow.com/questions/58293074/best-way-to-implement-basic-content-manager-permissions

**Expected behavior** 
The experience for `Content Managers` should be the same as it is for `App` modules that were previously added by `Admin` or `Super User`.  They should have the ability to `Choose App` and select the appropriate layout/view, and subsequently manage the content.  Those seem to work just fine for `Content Managers`, but they also need to be able to successfully add new `App` modules to the page and work with them.

**Instructions to Reproduce the Problem**
Follow the setup in the **Current behavior** section:

* Create `Content Managers` role
* Assign `Edit` rights at the DNN page level.
* Within DNN Extensions, mark the `App` module with `Can Deploy` rights for `Content Managers`.
* Add `Permission by Group / Role` in 2sxc.
* Add app permissions security rule for `Content Managers` RoleId with `Edit (Create, Read, Update, Delete)`.
* Login as user belonging to `Content Managers` role.
* Add `App` module to page for which the user has page `Edit` rights.
* See issue with not being able to `Choose App`.

**Why change the behavior?**
It currently seems to be a bug.

**Your environment**  
See below.

* **2sxc version(s):** 12.8.0/12.8.1 (tested on both)
* **Browser:** [all]
* **DNN:** [9.10.2] 
* **Language:** [any/all]

**Anything you would like to add**

**Comment by iJungleboy** [comment]
My guess is that somewhere during the last year our checks for edit-permissions became checks for Is-Admin. We'll have to review this, could take 2-3 weeks.

**Comment by david-poindexter** [comment]
Unfortunately, this is going to be a major issue for us with this one client, but I certainly understand - this is the way open-source goes.  If I could jump in and make a contribution to fix it, I would, but I am not sure where to even start.  😆

**Comment by jeremy-farrance** [comment]
I doubt its useful info, but I was curious and tested it... I can confirm this problem exists on 2sxc v11.20.

**Comment by david-poindexter** [comment]
I am having a hard time understanding why nobody else is running into this issue after all this time.  🤔

**Comment by iJungleboy** [comment]
@david-poindexter From our experience it's really hard to just give edit permissions so basically everybody ends up getting admin permissions. That's also why the special role exists so that admins can edit, but don't access the advanced UIs. 
We'll still look into this of course. 
Changing this stuff is fairly tricky, as it changes security behavior. So even if you would suggest a change, we would need to spend some time to be really sure we don't open something incorrect. 
If you really need to prioritize this, a sponsoring of 4h would do wonders ;).

**Comment by david-poindexter** [comment]
> @david-poindexter From our experience it's really hard to just give edit permissions so basically everybody ends up getting admin permissions. That's also why the special role exists so that admins can edit, but don't access the advanced UIs. We'll still look into this of course. Changing this stuff is fairly tricky, as it changes security behavior. So even if you would suggest a change, we would need to spend some time to be really sure we don't open something incorrect. If you really need to prioritize this, a sponsoring of 4h would do wonders ;).

@iJungleboy I'm not sure I fully understand the points above, but I trust you know what you are talking about.  Tell me what "sponsoring of 4h" looks like and that may be something we can swing for sure!

**Comment by BillyShelton** [comment]
For my use case, I was actually able to get it working as desired by giving the Content Manager edit permissions on the module but not on the page.  Once they have edit permissions they are able to remove items and reorder the list without issue.

**Comment by david-poindexter** [comment]
> For my use case, I was actually able to get it working as desired by giving the Content Manager edit permissions on the module but not on the page. Once they have edit permissions they are able to remove items and reorder the list without issue.

@BillyShelton this does not match our use case.  We need a non-admin / non-super-user to have Page edit rights so they can add new modules (2sxc App) to the page and then use that module.  Right now, they cannot.  😢

**Comment by BillyShelton** [comment]
@david-poindexter  Yeah, I realized that was a difference in our use cases & wanted to clarify that point. I think it would be important that any solution to your scenario still leave the developer in control of allowing or disallowing that roles ability to add new modules.

**Comment by iJungleboy** [comment]
@tvatavuk could you do some minimal research? it may be as simple as some web-api decorators which prevent the use of this endpoint.

**Comment by jeremy-farrance** [comment]
@iJungleboy - did you post the comment above on the wrong issue?

**Comment by iJungleboy** [comment]
@tvatavuk probably found the root cause. It seems that if no app is selected, some security checks are skipped, which would result in edit=true. We're looking into this.

**Comment by jeremy-farrance** [comment]
I was just sort of typing the same thing... will you also fix it in one of the v12 LTS versions?

**Comment by iJungleboy** [comment]
No. The effort to patch an older version is very big, with risks of other things slipping in. So we only do that for really critical stuff. 

I understand that this would be useful but it actually hasn't worked for a while and nobody noticed, so it won't make it into that category. + there are unsexy ways to live with it, so we won't invest the resources in v12.

Note that from our impression, v13 is very solid. So we'll probably release an LTS in 4-6 weeks.

**Comment by david-poindexter** [comment]
This is a very disappointing stance because regardless of whether or not anyone has noticed it, the feature is supposed to work, right?  And there is no LTS path forward at this time.  What is the purpose of LTS if there is no "long-term support"?  If it is money that is the barrier, please let us know how to sponsor this.  I received no response from my previous request about this.

There is no known workaround.  If I am wrong, please correct me.  We would LOVE to have a solution to this issue for the client!

**Comment by iJungleboy** [comment]
Hi David
The free support has no promises of course. 
We didn't get back to you on the sponsoring options before as we figured we would probably solve this anyhow. 
If you would be able to sponsor 4h to backport this to 12 LTS, please check https://2sxc.org/en/contact
Would be great ;)

**Comment by jeremy-farrance** [comment]
Based on what you have said, please consider not labelling versions LTS in the future. Free or not, LTS has a specific meaning and I recommend you just don't bother with the label again because it sets an expectation and not following through creates a serious "negative delta." 
 
I love 2sxc and I hope you take this input as professionally as it is intended. Thank you!!

**Comment by david-poindexter** [comment]
@iJungleboy with all due respect, the stated rates for open-source "sponsorship" are astronomical.  Nonetheless, we are presenting this as an option to our client for consideration.

**Comment by iJungleboy** [comment]
@jeremy-farrance The label has 2 important uses

1. We need to differentiate between fairly quick releases with minimal testing, and much more tested releases. 
2. For people who are low-risk, we want to be sure that we have some editions that also have security/stability security.

`LTS` was the best we could come up with, because "stable" would also be misleading, because it would make other releases appear unstable. 

We've explained this in the blog here: https://2sxc.org/en/blog/post/special-edition-2sxc-9-14-lts-long-term-support

Security and stability are the main concern. So ATM that's our best option. We're open to better suggestions. 

@david-poindexter My company sponsors development for more than 100k per year - which has amounted to over 1 Million USD in the last 10 years. That's real money that I personally shared with the world. I do this, because I love what I do and believe it makes a difference. 99.99% of all users don't pay me anything, and often get free support on StackOverflow + Tutorials + Docs which are unmatched in the Dnn Ecosystem. 

The 0.01% of people that do pay us something cause a lot of overhead till they actually do sponsor anything. And to be honest, those that do sponsor are then usually enterprises where the price is simply irrelevant. This is the logic behind the pricing of our support, I hope it does seem fair when looking at the big picture.

**Comment by david-poindexter** [comment]
@iJungleboy I have huge respect for anyone participating and supporting open-source initiatives.  That includes you and your team!  😉  

I also understand that you are leveraging this open-source project to actually generate revenue for your business, so although it is nice to hear the sentiment of such charity, it is not completely out of charity alone.  😉   Take a solution that works for your business and making it open-source is brilliant and we do the exact same thing with some of our open-source projects.  I guess we just have a different mindset on how we approach sponsored changes from community users of our open-source projects.  Because we use the tool to help us make money, it seems only fair (from our perspective) that we split the cost with the sponsor for changes that will provide us (and others) benefits from the proposed changes.  In practice, this has worked really well for us over the years and it helps us "partner" with our community to move things forward that are mutually beneficial, and without high development fees for our community.  On the flip side, sometimes we have clients that want changes to the underlying solution.  For those, we charge standard rates because it is a specific client request and billable accordingly.  This too has worked well for us over the years.

Open-source is not an exact science and we all have our own unique experiences.  So, to each his own I suppose.  I was just a bit shocked at the posture, for a BUG on an LTS version, and I have hardly ever asked for anything from this project.  Anyway, we'll see what our client says, but I doubt they are going to shell out $840 USD to fix something that they were expecting to work from the beginning.  Oh, and by the way, the hard sell for us is that the rate is higher than our agreed-upon rate for development in the first place.  So then we are in a position of potentially having to pay for something just to "save face" with our client.  It is a very uncomfortable position to be in, but I fully understand the risks of using open-source software solutions. 
 I just expected otherwise from such a seasoned project as 2sxc.  😢 

Regarding the use of `LTS`, I now better understand your dilemma.  Thanks for explaining.  I am pretty sure most outsiders will not view `LTS` in the same way though, so this may need to be thought through a bit more to come up with something that is clearer to users.  As @jeremy-farrance pointed out, `LTS` has a very specific meaning and expectations come with that meaning.

**Comment by iJungleboy** [comment]
We've added docs to explain what LTS means, and will link this in future LTS releases. 

👉 https://r.2sxc.org/lts

**Comment by david-poindexter** [comment]
![image](https://user-images.githubusercontent.com/4568451/154816939-e35691a8-96f8-48f1-b026-92425aef937c.png)

**Comment by jeremy-farrance** [comment]
I truly appreciate that all this was discussed publicly and to a reasonable conclusion. Very seriously, thank you Daniel - and thank you David!!

**Comment by iJungleboy** [comment]
@david-poindexter Thanks for being my first sponsor!

**Comment by david-poindexter** [comment]
@iJungleboy it's the least I can do and I am trying to help get more people sponsoring developers.  I talk about it every chance I get because I feel so many people still do not know there is a mechanism to do so.  By the way, the above was a screenshot from the DNN Community Slack workspace.  I highly recommend you join that as well.  2sxc is a topic of conversation multiple times per day.  ;-)  https://dnncommunity.slack.com

**Comment by iJungleboy** [comment]
@david-poindexter Just tried to join dnncommunity.slack.com, but it says 
![image](https://user-images.githubusercontent.com/5321186/154910645-ca098e84-1f99-48b3-b17d-0f2907b96d8c.png)

How would I join?

**Comment by iJungleboy** [comment]
I think I figured it out, using this link: https://github.com/dnnsoftware/Dnn.Platform/discussions/3714 

Hard to discover though, maybe there should be a better point of entry...

**Comment by david-poindexter** [comment]
Apologies @iJungleboy - here is a better link.  
https://join.slack.com/t/dnncommunity/shared_invite/zt-5oi1bydv-w0IAChDA3BxIVRsRs3fCDA
